pam_timestamp: replace hmac implementation

ikerexxe opened the pull request:
sha1 is no longer recommended as a cryptographic algorithm for
authentication. Thus, the idea of this change is to replace the
implementation provided by hmacsha1 included in pam_timestamp module by
the one in the openssl library. This way, there's no need to maintain
the cryptographic algorithm implementation and it can be easily changed
with a single configuration change.

modules/pam_timestamp/hmac_openssl_wrapper.c: implement wrapper
functions around openssl's hmac implementation. Moreover, manage the key
generation and its read and write in a file. Include an option to
configure the cryptographic algorithm in login.defs file.
modules/pam_timestamp/hmac_openssl_wrapper.h: likewise.
modules/pam_timestamp/pam_timestamp.c: replace calls to functions
provided by hmacsha1 by functions provided by openssl's wrapper.
configure.ac: include openssl dependecy if it is enabled.
modules/pam_timestamp/Makefile.am: include new files and openssl library
to compilation.
ci/install-dependencies.sh: include openssl library to dependencies.
NEWS: add new item to next release.

t8m commented on modules/pam_timestamp/Makefile.am:
```diff
@@ -32,15 +32,15 @@ endif
 sbin_PROGRAMS = pam_timestamp_check
 
-pam_timestamp_la_SOURCES = pam_timestamp.c hmacsha1.c sha1.c
+pam_timestamp_la_SOURCES = pam_timestamp.c hmacsha1.c sha1.c hmac_openssl_wrapper.c
```
I do not think you want to simultaneously build and link both the wrapper and the original implementation. It should be made conditional based on whether use of OpenSSL is enabled or not. And similarly below.

ikerexxe replied:
I've changed this part, please review it when possible.

t8m commented on modules/pam_timestamp/hmac_openssl_wrapper.c:
```diff
@@ -0,0 +1,405 @@
+#include "hmac_openssl_wrapper.h"
+
+#define LOGIN_DEFS          "/etc/login.defs"
+#define CRYPTO_KEY          "HMAC_CRYPTO_ALGO"
```
I am not really sure all this hmac algorithm configurability is needed here. If SHA2 based algorithms are broken then you do not have replacement among those supported here anyway. And if you really want to have algorithm agility, instead of hardcoding a table of algorithm names, you should use something like EVP_getdigestbyname() function to get the EVP_MD.

Also did you agree with shadow utils maintainer on addition of the new key value in login.defs?

ikerexxe replied:
> I am not really sure all this hmac algorithm configurability is needed here. If SHA2 based algorithms are broken then you do not have replacement among those supported here anyway. And if you really want to have algorithm agility, instead of hardcoding a table of algorithm names, you should use something like EVP_getdigestbyname() function to get the EVP_MD.

Awesome! I didn't know about this function. Thanks for pointing it out to me.

> Also did you agree with shadow utils maintainer on addition of the new key value in login.defs?

No, and this is something that I wanted to discuss but somehow I forgot about it. I doubt what's the best approach to this configuration. On the one hand there's the current approach. On the other hand it would be possible to create a `pam_timestamp.conf` file and include it there. Finally, it would be possible to reuse current `ENCRYPT_METHOD` in `login.defs`. What do you think is the best solution?

ikerexxe replied:
Can I get some feedback for this point? Every time I think about it I'm more willing to reuse `ENCRYPT_METHOD`as this  would mean changing only a single setting to improve the overall security of the system.

t8m replied:
The problem with ENCRYPT_METHOD potentially might be that there might be other values possible for ENCRYPT_METHOD in future that are not valid hash algorithm names. So basically you'd have to have some sensible default and if the ENCRYPT_METHOD fails to return valid hash algorithm then you'd have to fallback to the default.

ikerexxe replied:
Thank you for the analysis. Then I think I'll continue with the current approach, so I'll contact shadow-utils maintainers to get their opinion.

ikerexxe replied:
The key is available in shadow-utils master branch as https://github.com/shadow-maint/shadow/pull/333 got merged.

ikerexxe commented:
I have another doubt, this time concerning `hmac_size()` implementation. The current approach is agile from the development point of view as it leaves the actual details in the hands of openssl. The problem is that it takes more time to execute. On the other hand if I somehow hardcode the values then each time that an algorithm is changed the implementation would have to change. Do you think that the actual implementation is worth the execution overhead?

t8m commented:
I would not worry about the execution overhead.

ikerexxe commented on configure.ac:
```diff
@@ -642,6 +654,7 @@ AM_CONDITIONAL([COND_BUILD_PAM_SETQUOTA], [test "$ac_cv_func_quotactl" = yes])
 AM_CONDITIONAL([COND_BUILD_PAM_TTY_AUDIT], [test "$HAVE_AUDIT_TTY_STATUS" = yes])
 AM_CONDITIONAL([COND_BUILD_PAM_UNIX], [test "$enable_unix" = yes])
 AM_CONDITIONAL([COND_BUILD_PAM_USERDB], [test -n "$LIBDB"])
+AM_CONDITIONAL([COND_BUILD_OPENSSL], [test "x$use_openssl" = "xyes"])
```
Fixed

ikerexxe commented:
Hi @t8m,

Can I have a second round of reviews for this PR?

t8m commented:
> Can I have a second round of reviews for this PR?

I am sorry but I am unfortunately very busy with other things currently. I'll try to look at this in a week or so.

ikerexxe commented:
Thanks for the feedback @t8m, I wasn't aware that there was a new API that provided the HMAC functionality and that the old one was being deprecated. I've updated all the calls to use only the new API, and removed the old one as from my point of view it doesn't make any sense to add new code that will be removed soon.

By the way, I've created a COPR repository to test the change: https://copr.fedorainfracloud.org/coprs/ipedrosa/openssl_pam_timestamp/

t8m commented on modules/pam_timestamp/hmac_openssl_wrapper.c:
```diff
@@ -0,0 +1,395 @@
+                char *key, size_t key_length,
+                OSSL_PARAM subalg_param[], const void *text, size_t text_length)
+{
+    int ret = 0;
```
Please use `ret = PAM_AUTH_ERR` here and remove all these assignments in the error cases.

t8m commented on modules/pam_timestamp/hmac_openssl_wrapper.c:
```diff
@@ -0,0 +1,395 @@
+    int ret = 0;
+    EVP_MAC *evp_mac = NULL;
+    EVP_MAC_CTX *ctx = NULL;
+    unsigned char* hmac_message = NULL;
```
Nit: `unsigned char *hmac_message...`

t8m commented on modules/pam_timestamp/hmac_openssl_wrapper.c:
```diff
@@ -0,0 +1,395 @@
+int
+hmac_size(pam_handle_t *pamh, size_t *hmac_length)
+{
+    int ret = 0;
```
As above.

t8m commented on modules/pam_timestamp/hmac_openssl_wrapper.c:
```diff
@@ -0,0 +1,395 @@
+
+    if (config_value != NULL) {
+        subalg_param[0] = OSSL_PARAM_construct_utf8_string(OSSL_MAC_PARAM_DIGEST,
+                                                           (char *)config_value,
```
The config_value will be leaked as it is allocated and you're not freeing it after the subalg_param is no longer used. I would probably not try to construct the params in get_crypto_algorithm() but just return the value from here and call the construct (and later free) where the algorithm string is used. For the DEFAULT_ALGORITHM I'd just return strdup().

t8m commented on modules/pam_timestamp/hmac_openssl_wrapper.c:
```diff
@@ -0,0 +1,395 @@
+        subalg_param[0] = OSSL_PARAM_construct_utf8_string(OSSL_MAC_PARAM_DIGEST,
+                                                           (char *)DEFAULT_ALGORITHM,
+                                                           0);
+        pam_syslog(pamh, LOG_WARNING,
```
This should be LOG_DEBUG at most and it should be called only if the debug parameter of the PAM module is set.

t8m commented on modules/pam_timestamp/pam_timestamp.c:
```diff
@@ -449,14 +456,22 @@ pam_sm_authenticate(pam_handle_t *pamh, int flags, int argc, const char **argv)
 
+#ifdef WITH_OPENSSL
+		if (hmac_size(pamh, debug, &maclen)) {
+			pam_syslog(pamh, LOG_CRIT, "Unable to get hmac size");
```
I would drop this log as there will be already a log message from the hmac_size() call. Too much logging is not good.

t8m commented on modules/pam_timestamp/hmac_openssl_wrapper.c:
```diff
@@ -0,0 +1,395 @@
+    *out_length = hmac_length;
+    *out = malloc(*out_length);
+    if (*out == NULL) {
+        pam_syslog(pamh, LOG_ERR, "Not enough memory");
```
LOG_CRIT here.

t8m commented on modules/pam_timestamp/hmac_openssl_wrapper.c:
```diff
@@ -0,0 +1,395 @@
+
+    hmac_message = (unsigned char*)malloc(sizeof(unsigned char) * MAX_HMAC_LENGTH);
+    if (!hmac_message) {
+        pam_syslog(pamh, LOG_ERR, "Not enough memory");
```
LOG_CRIT here.

t8m commented on modules/pam_timestamp/hmac_openssl_wrapper.c:
```diff
@@ -0,0 +1,395 @@
+
+    tmp = malloc(st.st_size);
+    if (!tmp) {
+        pam_syslog(pamh, LOG_ERR, "Not enough memory");
```
LOG_CRIT here

t8m commented on modules/pam_timestamp/hmac_openssl_wrapper.c:
```diff
@@ -0,0 +1,395 @@
+
+    tmp = malloc(key_size);
+    if (!tmp) {
+        pam_syslog(pamh, LOG_ERR, "Not enough memory");
```
LOG_CRIT here

t8m commented on modules/pam_timestamp/hmac_openssl_wrapper.c:
```diff
@@ -0,0 +1,395 @@
+
+    if (count < key_size) {
+        pam_syslog(pamh, LOG_ERR, "Short read on random device");
+        memset(tmp, 0, key_size);
```
This is probably useless as these are just random bytes.

t8m commented on modules/pam_timestamp/pam_timestamp.c:
```diff
@@ -487,8 +502,18 @@ pam_sm_authenticate(pam_handle_t *pamh, int flags, int argc, const char **argv)
+#ifdef WITH_OPENSSL
+		if (hmac_generate(pamh, debug, &mac, &maclen, TIMESTAMPKEY,
+				  ROOT_USER, ROOT_GROUP, message, message_end - message)) {
+			pam_syslog(pamh, LOG_ERR, "Error generating hmac");
```
There should be already a syslog message from hmac_generate, shouldn't it? If so I'd drop this to not spam the syslog too much.

t8m commented on modules/pam_timestamp/pam_timestamp.c:
```diff
@@ -605,8 +630,17 @@ pam_sm_open_session(pam_handle_t *pamh, int flags UNUSED, int argc, const char *
 
+#ifdef WITH_OPENSSL
+	if (hmac_size(pamh, debug, &maclen)) {
+		pam_syslog(pamh, LOG_CRIT, "Unable to get hmac size");
```
As above.

t8m commented on modules/pam_timestamp/pam_timestamp.c:
```diff
@@ -621,15 +655,22 @@ pam_sm_open_session(pam_handle_t *pamh, int flags UNUSED, int argc, const char *
+#ifdef WITH_OPENSSL
+	if (hmac_generate(pamh, debug, &mac, &maclen, TIMESTAMPKEY,
+			  ROOT_USER, ROOT_GROUP, text, p - text)) {
+		pam_syslog(pamh, LOG_ERR, "Error calculating hmac");
```
As above.

t8m reviewed (changes requested):
A few final nits. @ldv-alt Could you please also look?

t8m commented:
Approved. @ldv-alt OK to merge this?

ldv-alt commented on modules/pam_timestamp/hmac_openssl_wrapper.c:
```diff
@@ -0,0 +1,394 @@
+    }
+
+    count = 0;
+    while (count < key_size) {
+        i = read(fd, tmp + count, key_size - count);
+        if ((i == 0) || (i == -1)) {
+            break;
+        }
+        count += i;
+    }
```
Consider using pam_modutil_read.

ldv-alt commented on modules/pam_timestamp/hmac_openssl_wrapper.c:
```diff
@@ -0,0 +1,394 @@
+    }
+
+    count = 0;
+    while (count < (size_t)st.st_size) {
+        i = read(fd, tmp + count, st.st_size - count);
+        if ((i == 0) || (i == -1)) {
+            break;
+        }
+        count += i;
+    }
```
Consider using pam_modutil_read.

ldv-alt commented on modules/pam_timestamp/hmac_openssl_wrapper.c:
```diff
@@ -0,0 +1,394 @@
+    }
+
+    count = 0;
+    while (count < text_length) {
+        i = write(fd, text + count, text_length - count);
+        if ((i == 0) || (i == -1)) {
+            break;
+        }
+        count += i;
+    }
```
Consider using pam_modutil_write.

ldv-alt reviewed (commented):
Consider using functions from pam_modutil_ioloop.c

ikerexxe commented:
> Consider using functions from pam_modutil_ioloop.c

Done. Thank you for your feedback.

ldv-alt commented on modules/pam_timestamp/hmac_openssl_wrapper.c:
```diff
@@ -0,0 +1,374 @@
+    close(fd);
+
+    if (bytes_written < text_length) {
+        pam_syslog(pamh, LOG_ERR, "Short write on random device");
```
Is `file_name` really a random device?

ikerexxe replied:
No, not really. I changed it for the `file_name`.

ldv-alt commented on modules/pam_timestamp/hmac_openssl_wrapper.c:
```diff
@@ -0,0 +1,374 @@
+    bytes_read = pam_modutil_read(fd, tmp, st.st_size);
+    close(fd);
+
+    if (bytes_read < (size_t)st.st_size) {
+        memset(tmp, 0, st.st_size);
+        free(tmp);
+        return PAM_AUTH_ERR;
+    }
```
Should some diagnostics be printed in case of short read, too?

ikerexxe replied:
Yes, I've just added them.

t8m commented on modules/pam_timestamp/hmac_openssl_wrapper.c:
```diff
@@ -0,0 +1,374 @@
+    }
+
+    tmp = malloc(st.st_size);
+    if (!tmp) {
```
This might produce misleading syslog if st.st_size is 0 actually. Please check that.

ikerexxe replied:
I've added a check in the previous lines. Please review it.

ldv-alt reviewed (approved):
OK, I suppose this could be merged now.

t8m commented:
Merged. Thank you for the contribution!
